docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md

```diff
@@ -0,0 +1,245 @@
+
+This module depends on `MBEDTLS_RUNTIME_DETECTION`. It provides cpu feature detection function and feature masks.
+
+files:
+- `library/runtime.h`
+- `library/runtime.c`
+
+I prefer implement the function as util macro in `runtime.h` like below.
+
+```c
+#if defined(MBEDTLS_RUNTIME_DETECTION)
+#if defined(MBEDTLS_ARCH_IS_ARM64)
+#define mbedtls_cpu_has_feature( flags ) mbedtls_cpu_has_feature_arm64( flags )
+#elif defined(MBEDTLS_ARCH_IS_AMD64)
+#define mbedtls_cpu_has_feature( flags ) mbedtls_cpu_has_feature_amd64( flags )
+#else
+#error "No Arch defined"
+#endif
+#else
+/* always enable */
+#define mbedtls_cpu_has_feature( flags ) 1
+#endif
+```
+
+##### X86 and AMD64
+
+x86 and x86_64 provide well defined `cpuid` instructions, just followup it and put needed mask into `runtime.h`
+
+##### Arm64
+
+This is very complex issue. Fortunately, sha512 already provides a perfect implementation. We can extend it.
+
+1. Feature mask should followup [linux elf hwcaps](https://github.com/torvalds/linux/blob/master/Documentation/arm64/elf_hwcaps.rst)
+1. Translation other OS into the definition of linux when first call.
+
+#### Algo Module
+
+This section will take AES set encrypt key function as example.
+
+The function that can be replace with hardware accceleration should looks like below
+
+```c
+/* Without runtime detection, this function will be optimized */
+static int aes_internal_setkey_enc(unsigned char *rk,
+                                   const unsigned char *key,
+                                   size_t bits)
+{
+
+#if defined(MBEDTLS_ARCH_IS_ARM64)
+#if defined(MBEDTLS_AES_ARM64_CRYPTO)
+    if (mbedtls_cpu_has_support(MBEDTLS_CPU_HAS_CRYPTO)) {
+        return mbedtls_aesce_setkey_enc(rk, key, bits);
+    }
+#endif
+#if defined(MBEDTLS_AES_ARM64_SVEAES)
+    if (mbedtls_cpu_has_support(MBEDTLS_CPU_HAS_SVEAES)) {
+        return mbedtls_aesce_setkey_enc(rk, key, bits);
+    }
+#endif
+#endif /* MBEDTLS_ARCH_IS_ARM64 */
+
+#if defined(MBEDTLS_AES_NO_ACCEL)
+    return aes_internal_setkey_enc_no_accel(rk, key, bits);
+#endif
+
+}
+```
+
+#### Hardware acceleration Modules
+
+The file should be includes
+1. Compiler compatible and cpu modifier flags check
+1. Enable needed cpu modifier flags for runtime detection
+
+NOTE: How to do it for MSVC? I did not found it.
+
+It looks like bellow.
+
+```c
+#if defined(MBEDTLS_ARCH_IS_ARM64)
+
+#if defined(MBEDTLS_RUNTIME_DETECTION)
+#if defined(__clang__)
+#   if __clang_major__ < 4
+#       error "A more recent Clang is required for MBEDTLS_AES_C"
+           /* Workaround https://reviews.llvm.org/D131064 . */
+#   elif __clang_major__ < 15
+#       define __ARM_FEATURE_CRYPTO 1
+#   endif
+#   pragma clang attribute push (__attribute__((target("crypto"))), apply_to=function)
+#   define MBEDTLS_POP_TARGET_PRAGMA
+#elif defined(__GNUC__)
+#   if __GNUC__ < 6
+#       error "A more recent GCC is required for MBEDTLS_AES_C"
+#   else
+#       pragma GCC target ("arch=armv8-a+crypto")
```

Comment:
With `runtime detection`, I don't think so. 

This piece should be put into hardware acceleration module. In build scripts, if it is not supported, the module should be disabled. 

`#       error "A more recent GCC is required for MBEDTLS_AES_C"` should be changed to `#       error "A more recent GCC is required for MBEDTLS_AESCE_C"` . That might be more clearly

Reply:
On second thoughts, this may be the right thing. We need to tell the compiler that the accelerated AES code, and _only the accelerated AES code_, can use the crypto instructions. So this needs to be only enabled around the accelerated AES code, and not in a header file.

When I first read this section, I thought this snippet was supposed to be in a header file. Reading this again, I think you might have intended it to be in the C file? Please clarify.

Reply:
Sure. I will clarify it.

Reply:
it is removed